### PR TITLE
Add `torch.cuda.streams.ExternalStream`

### PR DESCRIPTION
**Pull request by emcastillo**
This is required in https://github.com/pytorch/pytorch/pull/57110#issuecomment-828357947

We need to provide means to synchronize on externally allocated streams for dlpack support in python array data api.

cc @mruberry @rgommers @leofang @asi1024 @kmaehashi

### CODE REVIEW

**Comment by leofang** [comment]
cc: @gmarkall @pentschev @jakirkham @kkraus14 for vis

File: c10/cuda/CUDAStream.cpp [hunk 1]
```diff
@@ -64,10 +64,13 @@ static LeakyStreamInternals default_streams[C10_COMPILE_TIME_MAX_GPUS];
 static std::once_flag device_flags[C10_COMPILE_TIME_MAX_GPUS];
 static std::atomic<uint32_t> low_priority_counters[C10_COMPILE_TIME_MAX_GPUS];
 static std::atomic<uint32_t> high_priority_counters[C10_COMPILE_TIME_MAX_GPUS];
+static std::atomic<uint32_t> external_counters[C10_COMPILE_TIME_MAX_GPUS];
 static std::array<LeakyStreamInternals, kStreamsPerPool>
     low_priority_streams[C10_COMPILE_TIME_MAX_GPUS];
 static std::array<LeakyStreamInternals, kStreamsPerPool>
     high_priority_streams[C10_COMPILE_TIME_MAX_GPUS];
+static std::array<LeakyStreamInternals, kStreamsPerPool>
```
Comment (emcastillo) [inline c10/cuda/CUDAStream.cpp:72]: I am not fond of this, as in the current implementation it causes stream reusing when 32 streams are created on the same device without the user knowing, so the stream pools must be rewritten, or streams should be created and freed in an RAII way. However this should be all addressed together in a different PR.

Comment (leofang) [reply c10/cuda/CUDAStream.cpp:72]: You should open a new issue with you nice wtf bug reproducer 😂

**Comment by ngimel** [comment]
See previous attempt https://github.com/pytorch/pytorch/pull/39567

File: c10/cuda/CUDAStream.cpp (hunk 1 above)
Comment (ezyang) [reply c10/cuda/CUDAStream.cpp:72]: It's pretty fun because this patch has the *exact* same limitations as the previous version of the patch LOL

File: c10/cuda/CUDAStream.cpp [hunk 2]
```diff
@@ -100,6 +104,7 @@ enum class StreamIdType : uint8_t {
   DEFAULT = 0x0,
   LOW = 0x1,
   HIGH = 0x2,
+  EXT = 0x4,
```
Comment (ezyang) [inline c10/cuda/CUDAStream.cpp:107]: The bit numbering here is a misunderstanding of what we need. This is not a bitmask. You can set EXT as 0x3.

File: torch/cuda/streams.py [hunk 3]
```diff
@@ -112,6 +112,29 @@ def __repr__(self):
...
+
+    Args:
+        stream_ptr(int): Integer representation of the `cudaStream_t` value.
+            allocated externally.
+        device(torch.device or int, optional): the device where the stream
+            was originally allocated.
+    """
+
+    def __new__(cls, stream_ptr, device=None, **kwargs):
+        print('Creating new stream ', stream_ptr)
```
Comment (ezyang) [inline torch/cuda/streams.py:133]: delete me

File: torch/cuda/streams.py [hunk 4]
```diff
@@ -112,6 +112,29 @@ def __repr__(self):
...
 
+class ExternalStream(Stream):
+    r"""Wrapper around a externally allocated CUDA stream.
+
+    A CUDA stream is a linear sequence of execution that belongs to a specific
+    device, independent from other streams.  See :ref:`cuda-semantics` for
+    details.
+
+    This class is used to wrap streams allocated in other libraries in order
+    to facilitate data exchange and multi-library interactions.
```
Comment (ezyang) [inline torch/cuda/streams.py:123]: Should have a warning saying that only a fixed number of external streams per device are OK, then you ded

File: c10/cuda/CUDAStream.cpp (hunk 1 above)
Comment (ngimel) [reply c10/cuda/CUDAStream.cpp:72]: Is there a real bug here? 32 streams in pool limitation was chosen because hardware streams are limited in number, so even though, technically, you can create unbounded number of user facing soft streams, they'd still be backed by the same hardware streams.

Comment (leofang) [reply c10/cuda/CUDAStream.cpp:72]: I think there are a few catches here. Disclaimer is I haven't looked into the detail of PyTorch's stream pool, so I am just applying common sense here 😅 

>> 32 streams in pool limitation was chosen because hardware streams are limited in number, so even though, technically, you can create unbounded number of user facing soft streams, they'd still be backed by the same hardware streams.
>
>  we expect there to be fewer than 32 actual hardware streams. 
1. I think the number of hardware streams is an implementation detail in CUDA/HIP's runtime, and users should not need to know nor would they notice (unless concurrency is a concern). But the current implementation seems to expose it in an unexpected way -- @emcastillo can say more about it.
2. A CUDA/HIP user always expects that if `cudaStreamCreate()` succeeds then the returned stream is valid. On a visual profiler (`nvvp` or `nsys`) they will see work done on each designated stream (the stream number can go far beyond 32), which reveals no information on the actual number of hardware streams, as expected.
3. 32 is an outdated number. On recent hardware it is at least 128 (see the first row in [this table](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications__technical-specifications-per-compute-capability)). But, again, it's implementation detail (and hardware dependent). Is there a good argument to hard-code it?

I am simply curious about the design choice and not really pushing for any change (either in this PR or a separate one).

Comment (emcastillo) [reply c10/cuda/CUDAStream.cpp:72]: > One hacky temporary fix would be to prohibit more external streams than the cache size so there's never premature eviction. But a more durable fix would probably be to modify the CUDAStream class or {super, sub}class it to permit external streams and avoid a "pool" or cache. @ngimel and I don't think we want to modify PyTorch's existing stream pools, which have a very different usage pattern, and were designed to avoid constant stream creation and destruction. In this case, however, we do want the external streams to have a temporary lifetime.

I totally agree with this point. I will try to subclass it again,
I remember trying to do it originally but found some issues that I can't remember (most likely due to my own ineptitude)

Comment (mruberry) [reply c10/cuda/CUDAStream.cpp:72]: > 1. I think the number of hardware streams is an implementation detail in CUDA/HIP's runtime, and users should not need to know nor would they notice (unless concurrency is a concern). But the current implementation seems to expose it in an unexpected way -- @emcastillo can say more about it.

It is an implementation detail and this does not expose it. If a user does use 32+ user-facing streams then they could observe the stream pool round-robining the pool streams, however. We were concerned that this might confuse some users but in practice it doesn't seem to be an issue.

> 2. A CUDA/HIP user always expects that if `cudaStreamCreate()` succeeds then the returned stream is valid. On a visual profiler (`nvvp` or `nsys`) they will see work done on each designated stream (the stream number can go far beyond 32), which reveals no information on the actual number of hardware streams, as expected.
> 3. 32 is an outdated number. On recent hardware it is at least 128 (see the first row in [this table](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications__technical-specifications-per-compute-capability)). But, again, it's implementation detail (and hardware dependent). Is there a good argument to hard-code it?

The returned stream will always be valid. It's also important to note that the pool supports accessing 32 user-facing hardware streams but libraries like cuBLAS and cuDNN independently manage their own streams. If there's a real program suffering because PyTorch only supports 32 user-facing streams hardware streams then we could consider increasing the pool size.

> I am simply curious about the design choice and not really pushing for any change (either in this PR or a separate one).

The value might be outdated and we could consider increasing it on newer hardware, but we haven't heard that it's an issue. We could also look at alternative designs where streams were created as usual but then returned to the pool when they were about to be destroyed for reuse later.

**Comment by ezyang** [comment]
According to @ngimel the only blocker is making sure you don't silently overflow the 32 external streams you get (just need to add a little error checking there)

**Comment by emcastillo** [comment]
@ezyang tahnks for the follow up!

> don't silently overflow the 32 external streams you get (just need to add a little error checking there)

I believe this is not as easy as it might seem.
Once we get to assign a stream, we can't currently release it and mark it available on the pool. Currently, the pool overwrites older streams.

```python

import torch
streams = []
stream_ids = set()
for i in range(128):
    streams.append(torch.cuda.Stream())
    if streams[-1].cuda_stream in stream_ids:
        raise ValueError(f'Stream {i} reallocated')
    else:
        stream_ids.add(streams[-1].cuda_stream)
----

Traceback (most recent call last):
  File "stream_stress.py", line 8, in <module>
    raise ValueError(f'Stream {i} reallocated')
ValueError: Stream 32 reallocated
```

There is no destructor in CUDAStream that allows to mark a stream to be used or unused so I believe the only cleaner way is to have a hash map that holds the externally allocated streams with the `LeakyStreamInternals` instead of a "pool" for external streams, and subclass CUDAStream that manages this map by filling it when a stream is requested, and removing it when the subclass objects are deleted. How do you think?

**Comment by dzhulgakov** [comment]
Have we considered doing pointer bit packing tricks? :) Stream/CudaStream are currently 1 byte for device type, 1 byte for device index and 4 bytes for stream id: https://github.com/pytorch/pytorch/blob/master/c10/core/Stream.h#L120

We don't seem to pass around raw stream id much at all. So let's say that we extend StreamId to 6 bytes. On x86 pointers are only 48 bits long, so we can easily pack the cuda_stream_t pointer as the cuda_stream_id. Of course we need to distinguish stream type, but it's easy, currently CUDA stream has the following scheme allocating the ids: https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDAStream.cpp#L76 - they are always small (7 bits) so those aren't valid pointers either. This way we can just assume that if StreamId is more than say 256 - it's reinterpreted as cudaStream_t (which is defined as a pointer btw).

I know it sounds crazy, but it's a small change from the today's code and doesn't bring any need for the bookkeeping / static limitations

**Comment by ezyang** [comment]
Bit packing is very reasonable and might just moot our entire conversation here. However, just to clarify the earlier point:

@emcastillo Sorry, I wasn't clear, I was suggesting something very very dumb: just hard error at the point we overflow. Don't bother trying to find unused stream ids that can be reused. That should be simple to implement, and blow up very loudly when someone tries to use too many external streams.

**Comment by dzhulgakov** [comment]
> Similarly the CUDAStream class would get a destructor that would check if the stream was "external" and, if so, destroy the stream?

We can go even simpler and say that the ownership of the external stream stays with the caller. So we don't even need the destructor

**Comment by emcastillo** [comment]
I think it is reasonable to assume that the stream life-cycle is responsibly of the caller.
Thanks @dzhulgakov! I will try to implement the bitpacking trick :D

**Comment by leofang** [comment]
In the DLPack protocol, the consumer hands its stream to the producer, and the producer decides what to do (to sync immediately or just establish the stream order).

**Comment by emcastillo** [comment]
pushed the @dzhulgakov suggestion!
@ezyang @mruberry can you take a look :)?

File: torch/csrc/cuda/Stream.cpp [hunk 5]
```diff
@@ -38,8 +39,10 @@ static PyObject * THCPStream_pynew(
   at::cuda::CUDAStream stream =
     cdata ?
     at::cuda::CUDAStream::unpack(cdata) :
-    at::cuda::getStreamFromPool(
-      /* isHighPriority */ priority < 0 ? true : false);
+      stream_ptr ?
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply torch/csrc/cuda/Stream.cpp:46]: Good idea! :)

**Comment by ezyang** [comment]
This new version looks extremely promising. I probably only have some quibbling with the safety of pointer tagging as it is implemented here. Maybe we can just land this as is.

File: c10/core/Stream.h [hunk 6]
```diff
@@ -119,21 +119,21 @@ class Stream final {
...
     static_assert(sizeof(DeviceIndex) == 1, "DeviceIndex is not 8-bit");
-    static_assert(sizeof(StreamId) == 4, "DeviceIndex is not 32-bit");
+    static_assert(sizeof(StreamId) == 8, "DeviceIndex is not 64-bit");
     // Concat these together into a 64-bit integer
     // See Note [Hazard when concatenating signed integers]
     uint64_t bits = static_cast<uint64_t>(static_cast<uint8_t>(device_type()))
-            << 48 |
-        static_cast<uint64_t>(static_cast<uint8_t>(device_index())) << 32 |
+            << 56 |
+        static_cast<uint64_t>(static_cast<uint8_t>(device_index())) << 48 |
```
Comment (ezyang) [inline c10/core/Stream.h:127]: It would be nice if there was an assert here that the bits we masked out were the signed extended highest order bit of the valid 48 bit segment, would make this code marginally more future proof.

Comment (ezyang) [reply c10/core/Stream.h:127]: Actually, the code here is technically wrong. Because addresses are sign extended from the valid 48-bit range to 64-bit range, you are obligated to mask out the top 16 bits (and then when unpacking, sign extend the 48 bits). If you are given 0xFFFFFFFF FFFFFFFF (which is a valid address even under 48-bit assumption) you'd clobber the device type and device index here.

File: c10/core/Stream.h [hunk 7]
```diff
@@ -119,21 +119,21 @@ class Stream final {
...
         static_cast<uint64_t>(static_cast<uint32_t>(id()));
     return bits;
   }
 
   static Stream unpack(uint64_t bits) {
-    const auto stream_id = static_cast<StreamId>(bits & 0xFFFFFFFFull);
-    bits >>= 32;
+    const auto stream_id = static_cast<StreamId>(bits & 0xFFFFFFFFFFFFull);
+    bits >>= 48;
```
Comment (ezyang) [inline c10/core/Stream.h:134]: sign extension needed here

File: c10/cuda/CUDAStream.cpp [hunk 8]
```diff
@@ -125,7 +130,13 @@ std::ostream& operator<<(std::ostream& stream, StreamIdType s) {
 // see Note [Hazard when concatenating signed integers]
 
 static inline StreamIdType streamIdType(StreamId s) {
-  return static_cast<StreamIdType>(s >> kStreamsPerPoolBits);
+  if (s > 256) {
```
Comment (ezyang) [inline c10/cuda/CUDAStream.cpp:133]: Hard coded magic number makes me nervous. More shortly.

File: c10/cuda/CUDAStream.cpp [hunk 9]
```diff
@@ -77,9 +77,10 @@ static std::array<LeakyStreamInternals, kStreamsPerPool>
 // zeros         StreamIdType  stream id index
```
Comment (ezyang) [inline c10/cuda/CUDAStream.cpp:77]: The diagram here has to be updated right

**Comment by ezyang** [comment]
What makes me a little nervous about the current stream id allocation is that you are using specifying pointer values below 256 to be hard coded stream IDs (in no small part because the null ptr aka zero has to denote the default stream id), but all of these values are technically legal pointers. This is incredibly unlikely to be a problem in practice, as suggested by https://stackoverflow.com/questions/9249619/is-the-pointer-guaranteed-to-be-a-certain-value because most operating systems have conventions that prevent heap allocated addresses from living here (e.g., because it's being used for stack space, or because it's mapped as null equivalent), but it still makes me nervous.

It seems to me like a safer scheme would be to use the least significant bits (we get 3 bits due to 8 byte alignment) to indicate if something is a pooled stream. (For external readers: we can't use the 16 most significant bits because they are already being used to pack device type and device id in packed format--we could choose not to do this, but it would increase the blast radius as we'd have to change how Python Stream bindings work.) So the new stream ID layout would be this (only showing 16 bits but it extends in the obvious way):

* `0b 0000 0000 0000 0000` = default stream (nullptr)
* `0b 0000 0XXX XXXX X001` = high priority streams
* `0b 0000 0XXX XXXX X010` = low priority streams
* `0b YYYY YYYY YYYY Y000` = external streams

What do you think?

File: c10/core/Stream.h (hunk 6 above)
Comment (emcastillo) [reply c10/core/Stream.h:127]: lol, you are absolutely right and made me ashamed of myself 😂

**Comment by emcastillo** [comment]
Updated!
@ezyang thanks for the awesome suggestions! I reworked the Ids as you suggested and I think it is much better now :).

File: test/test_cuda.py [hunk 10]
```diff
@@ -1314,6 +1314,20 @@ def test_record_stream_on_shifted_view(self):
 
         self.assertNotEqual(try_realloc.data_ptr(), data_ptr)
 
+    def test_external_streams(self):
+        stream_a = torch.cuda.Stream()
+        ext_stream = torch.cuda.streams.ExternalStream(stream_a.cuda_stream)
+        self.assertEqual(stream_a.cuda_stream, ext_stream.cuda_stream)
+
+    @unittest.skipIf(not TEST_MULTIGPU, "detected only one GPU")
```
Comment (mruberry) [inline test/test_cuda.py:1322]: cc @ptrblck: not for this PR, of course, but as a note for us it'd be nice to update test_cuda.py to use the device generic test suite, which has decorators for multiple devices and supports options like running the test suite on the first CUDA device

File: torch/cuda/streams.py [hunk 11]
```diff
@@ -112,6 +112,29 @@ def __repr__(self):
                 .format(self.device, self.cuda_stream))
 
 
+class ExternalStream(Stream):
+    r"""Wrapper around a externally allocated CUDA stream.
```
Comment (mruberry) [inline torch/cuda/streams.py:116]: "a" -> "an"

File: torch/cuda/streams.py [hunk 12]
```diff
@@ -112,6 +112,29 @@ def __repr__(self):
...
+
+    A CUDA stream is a linear sequence of execution that belongs to a specific
+    device, independent from other streams.  See :ref:`cuda-semantics` for
+    details.
+
+    This class is used to wrap streams allocated in other libraries in order
+    to facilitate data exchange and multi-library interactions. Be aware
+    that the stream life-cycle is managed by the owner.
```
Comment (mruberry) [inline torch/cuda/streams.py:124]: This is a nice explicit mention of the stream's life cycle, but I think we might want to be even clearer here. 

Maybe this could explain that the ExternalStream object doesn't keep the underlying stream alive, and that the user is expected to keep the original stream alive as long as the external stream object is in scope / alive?

**Review by mruberry — COMMENTED** [review]
I made two small suggestions for the docs. The tests look OK, and the approach looks much better!

File: c10/core/Stream.h [hunk 13]
```diff
@@ -119,21 +119,32 @@ class Stream final {
...
-            << 48 |
-        static_cast<uint64_t>(static_cast<uint8_t>(device_index())) << 32 |
-        static_cast<uint64_t>(static_cast<uint32_t>(id()));
+            << 56 |
+        static_cast<uint64_t>(static_cast<uint8_t>(device_index())) << 48 |
+        // Remove the sign extension part of the 64-bit address because
+        // the id might be used to hold a pointer.
+        (static_cast<uint64_t>(id()) & ((1ULL << 48) - 1));
+    TORCH_INTERNAL_ASSERT(
+        static_cast<DeviceIndex>((bits >> 48) & 0xFFFFull) == device_index(),
```
Comment (ngimel) [inline c10/core/Stream.h:132]: why is device_index masked with FFFF? It's only 8 bits (49 to 56) now, so the mask should be FF? (It used to be FFFF, when device_index was from bits 33 to 48).

File: c10/cuda/CUDAStream.cpp [hunk 14]
```diff
@@ -113,28 +120,39 @@ std::ostream& operator<<(std::ostream& stream, StreamIdType s) {
...
+// StreamId is 64-bit, so we can just rely on regular promotion rules.
 // We rely on streamIdIndex and streamIdType being non-negative;
 // see Note [Hazard when concatenating signed integers]
 
 static inline StreamIdType streamIdType(StreamId s) {
-  return static_cast<StreamIdType>(s >> kStreamsPerPoolBits);
+  int mask_for_type = (1 << kStreamTypeBits) - 1;
+  if (s && ((s & mask_for_type) == 0)) {
+    // Externally allocated streams have their id being the cudaStream_ptr
+    // so the bits ocrresponding to the type will be 0 and will collide with
```
Comment (ngimel) [inline c10/cuda/CUDAStream.cpp:141]: corresponding

File: c10/cuda/CUDAStream.cpp [hunk 15]
```diff
@@ -305,9 +323,19 @@ CUDAStream CUDAStream_fromInternals(const LeakyStreamInternals* ptr) {
...
+  // the stream_ids managed from the pool have only 8 bits
+  // so any value higher than that can be reinterpreted
+  // as an externally allocated stream
+  if (streamIdType(stream_id) == StreamIdType::EXT) {
+    // In this case this is a externally allocated stream
+    // we don't need to manage its life cycle
+    return reinterpret_cast<cudaStream_t>(stream_id);
+  } else {
+    auto ptr = CUDAStream_internals(*this);
+    AT_ASSERT(ptr);
```
Comment (ngimel) [inline c10/cuda/CUDAStream.cpp:336]: TORCH_CUDA_ASSERT

File: torch/cuda/streams.py [hunk 16]
```diff
@@ -112,6 +112,29 @@ def __repr__(self):
                 .format(self.device, self.cuda_stream))
 
 
+class ExternalStream(Stream):
+    r"""Wrapper around a externally allocated CUDA stream.
+
+    A CUDA stream is a linear sequence of execution that belongs to a specific
```
Comment (ngimel) [inline torch/cuda/streams.py:118]: Let's avoid circular references "independent from other streams"
A CUDA stream is a sequence of operations that execute in order on a specific device.

File: test/test_cuda.py [hunk 17]
```diff
@@ -1314,6 +1314,20 @@ def test_record_stream_on_shifted_view(self):
...
+    def test_external_streams_multi_device(self):
+        device = torch.device('cuda:1')
+        stream_a = torch.cuda.Stream(device=device)
```
Comment (ngimel) [inline test/test_cuda.py:1325]: do you want to have tests with real external streams, and not native streams masquerading as external?
```
lib = ctypes.cdll.LoadLibrary(None)
p=ctypes.c_void_p()
out=lib.cudaStreamCreate(ctypes.byref(p))
```

File: torch/cuda/streams.py [hunk 18]
```diff
@@ -112,6 +112,29 @@ def __repr__(self):
...
+    details.
+
+    This class is used to wrap streams allocated in other libraries in order
+    to facilitate data exchange and multi-library interactions. Be aware
+    that the stream life-cycle is managed by the owner.
+
+    Args:
+        stream_ptr(int): Integer representation of the `cudaStream_t` value.
+            allocated externally.
+        device(torch.device or int, optional): the device where the stream
```
Comment (ngimel) [inline torch/cuda/streams.py:129]: we should mention that if device was specified incorrectly, subsequent launches using this stream may fail. Unfortunately, there's no way to verify if device is correct or not.

File: test/test_cuda.py (hunk 17 above)
Comment (emcastillo) [reply test/test_cuda.py:1325]: this is pretty cool, thanks!!!

File: torch/cuda/streams.py (hunk 12 above)
Comment (emcastillo) [reply torch/cuda/streams.py:124]: Changed to the following paragrah

```
   This class doesn't manage the stream life-cycle, it is the user
    responsibility to keep the referenced stream alive while this class is
    being used.
```

File: torch/cuda/streams.py (hunk 16 above)
Comment (emcastillo) [reply torch/cuda/streams.py:118]: I just deleted that comment and replaced the entire docstring by
```
r"""Wrapper around an externally allocated CUDA stream.

    This class is used to wrap streams allocated in other libraries in order
    to facilitate data exchange and multi-library interactions.

    This class doesn't manage the stream life-cycle, it is the user
    responsibility to keep the referenced stream alive while this class is
    being used.

    Args:
...
```

How do you think?
Thanks!

File: torch/cuda/streams.py (hunk 18 above)
Comment (emcastillo) [reply torch/cuda/streams.py:129]: noted!

File: c10/cuda/CUDAStream.cpp (hunk 15 above)
Comment (emcastillo) [reply c10/cuda/CUDAStream.cpp:336]: changed the AT_ASSERTS in the file to TORCH_CUDA_ASSERT

Comment (emcastillo) [reply c10/cuda/CUDAStream.cpp:336]:
```
error: ‘TORCH_CUDA_ASSERT’ was not declared in this scope
   TORCH_CUDA_ASSERT(ptr);
```

actually, there is no `TORCH_CUDA_ASSERT` in the codebase 😅 I grepped it, but didn't find any :(

Comment (ngimel) [reply c10/cuda/CUDAStream.cpp:336]: Sorry, TORCH_INTERNAL_ASSERT. AT_ASSERT is deprecated

**Comment by emcastillo** [comment]
Fixed the assert and tests as @ngimel suggested :)

File: c10/core/Stream.h [hunk 19]
```diff
@@ -119,21 +119,32 @@ class Stream final {
...
+        static_cast<uint64_t>(static_cast<uint8_t>(device_index())) << 48 |
+        // Remove the sign extension part of the 64-bit address because
+        // the id might be used to hold a pointer.
+        (static_cast<uint64_t>(id()) & ((1ULL << 48) - 1));
+    TORCH_INTERNAL_ASSERT(
+        static_cast<DeviceIndex>((bits >> 48) & 0xFFull) == device_index(),
+        "DeviceIndex is not correctly packed");
+    TORCH_INTERNAL_ASSERT(
+        static_cast<DeviceType>((bits >> 56)) == device_type(),
+        "DeviceType is not correctly packed");
```
Comment (ezyang) [inline c10/core/Stream.h:136]: Actually, I feel like testing that id() is correctly packed is the most important one 🤣

File: c10/core/Stream.h [hunk 20]
```diff
@@ -119,21 +119,32 @@ class Stream final {
...
 
   static Stream unpack(uint64_t bits) {
-    const auto stream_id = static_cast<StreamId>(bits & 0xFFFFFFFFull);
-    bits >>= 32;
+    // Re-extend the sign of stream_id
+    uint64_t mask = (1ULL << 47);
+    const auto stream_id =
+        (static_cast<StreamId>(bits & 0xFFFFFFFFFFFFull) ^ mask) - mask;
+    bits >>= 48;
     const auto device_index = static_cast<DeviceIndex>(bits & 0xFFFFull);
```
Comment (ngimel) [inline c10/core/Stream.h:146]: Per comment above, mask here should also be FF?

Comment (emcastillo) [reply c10/core/Stream.h:146]: yeah, forget to change it here ... :(
Thanks!

File: c10/core/Stream.h (hunk 19 above)
Comment (emcastillo) [reply c10/core/Stream.h:136]: 🤣 will check!

**Comment by ezyang** [comment]
I'm slowly grinding through the bit patterns

**Comment by ezyang** [comment]
I verified that the bit twiddling correctly works in the sign extended case using [Crux](https://crux.galois.com/index.html) (thanks @atomb) using the following program:

```
#include <stdint.h>
#include <crucible.h>

uint64_t pack(int8_t device_index, int8_t device_type, int64_t id) {
  uint64_t bits = (uint64_t)((uint8_t)(device_type))
          << 56 |
      (uint64_t)((uint8_t)(device_index)) << 48 |
      // Remove the sign extension part of the 64-bit address because
      // the id might be used to hold a pointer.
      ((uint64_t)(id) & ((1ULL << 48) - 1));
  return bits;
}

int main() {
  int8_t device_index = crucible_int8_t("device_index");
  int8_t device_type = crucible_int8_t("device_type");
  int64_t id = crucible_int64_t("id");

  for (int i = 47; i < 64; i++) {
    assuming(((uint64_t)id) >> i & 1 == 1);
  }

  uint64_t bits = pack(device_index, device_type, id);

  // Re-extend the sign of stream_id
  uint64_t mask = (1ULL << 47);
  int64_t out_id =
      ((int64_t)(bits & 0xFFFFFFFFFFFFull) ^ mask) - mask;
  bits >>= 48;
  int8_t out_device_index = (int8_t)(bits & 0xFFFFull);
  bits >>= 8;
  int8_t out_device_type = (int8_t)(bits);

  check(device_index == out_device_index);
  check(device_type == out_device_type);
  check(id == out_id);
}
```

File: c10/cuda/CUDAStream.cpp [hunk 21]
```diff
@@ -305,9 +323,19 @@ CUDAStream CUDAStream_fromInternals(const LeakyStreamInternals* ptr) {
 } // anonymous namespace
 
 cudaStream_t CUDAStream::stream() const {
-  auto ptr = CUDAStream_internals(*this);
-  AT_ASSERT(ptr);
-  return ptr->stream;
+  int64_t stream_id = unwrap().id();
+  // the stream_ids managed from the pool have only 8 bits
```
Comment (ezyang) [inline c10/cuda/CUDAStream.cpp:327]: this comment out of date now

**Review by ezyang — APPROVED** [review]
this is so great, thank you so much for writing this

**Comment by ezyang** [comment]
need to skip tests on rocm

**Comment by ezyang** [comment]
Needs internal updates (related to int to int64)

**Comment by ezyang** [comment]
needs more internal updates (related to CUDA)

**Comment by emcastillo** [comment]
what happened? 😥

**Comment by ngimel** [comment]
Windows doesn't like ctypes https://app.circleci.com/pipelines/github/pytorch/pytorch/331129/workflows/3ecc5f8a-e782-4718-beca-1aea7a7433ce/jobs/13909012

**Comment by ezyang** [comment]
@emcastillo If you could push a new copy of the PR that would be a help. My guess is the use of ctypes to access CUDA library functionality is what is not working. We can instead stick all the functions we need in torch/csrc/cuda/shared/cudart.cpp and that should be good enough to get the test going.

**Comment by emcastillo** [comment]
> @emcastillo If you could push a new copy of the PR that would be a help. My guess is the use of ctypes to access CUDA library functionality is what is not working. We can instead stick all the functions we need in torch/csrc/cuda/shared/cudart.cpp and that should be good enough to get the test going.

Sure! let me work on it.
